### PR TITLE
gemini - websockets

### PR DESCRIPTION
**Pull request by pcriadoperez**
- For `fetchOHLCV` - allow sending both unified and exchange timeframe
- Add Gemini websockets

Collaboration with @carlosmiei 

@carlosmiei , I think I figured out why it wasn't working for you last time we looked at this. For some reason when using the marketId sometimes it works and sometimes it returns an invalid symbol error, sending in the exact same symbol. But if I upperCase the marketId it seems to consistently work. So hopefully now it works for you correctly.

Also added the ohlcv endpoint that wasn't there before.

### CODE REVIEW

**Comment by carlosmiei** [comment]
> image not sure if this pr is ready or i did wrong, but i tried using this pr, this line did not get transpiled properly to python

@leastchaos going to review it soon, thanks for the warning.

File: js/pro/gemini.js [hunk 1]
```diff
@@ -0,0 +1,663 @@
...
+        }
+        let stored = this.safeValue (this.ohlcvs[symbol], timeframe);
+        if (stored === undefined) {
+            const limit = this.safeInteger (this.options, 'OHLCVLimit', 1000);
+            stored = new ArrayCacheByTimestamp (limit);
+            this.ohlcvs[symbol][timeframe] = stored;
+        }
+        const changesLength = changes.length;
+        // reverse order of array to store candles in ascending order
+        for (let i = changesLength - 1; i >= 0; i--) {
```
Comment (carlosmiei) [inline js/pro/gemini.js:275]: yep this is not supported, but I think we can easily do this:
```Javascript
        for (let i = 0; i < changesLength; i++) {
            const index = changesLength - i;
            const parsed = this.parseOHLCV (changes[index], market);
            stored.append (parsed);
        }
```

File: js/pro/gemini.js [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+        const market = this.safeMarket (marketId);
+        const symbol = market['symbol'];
+        const messageHash = 'orderbook:' + symbol;
+        let orderbook = this.safeValue (this.orderbooks, symbol);
+        if (orderbook === undefined) {
+            orderbook = this.orderBook ();
+        }
+        for (let i = 0; i < changes.length; i++) {
+            const delta = changes[i];
+            const price = this.safeFloat (delta, 1);
```
Comment (carlosmiei) [inline js/pro/gemini.js:328]: @pcriadoperez btw we should avoid using `safeFloat`because it does not take into consideration the precision mode, `safeNumber` is preferred

File: js/pro/gemini.js [hunk 3]
```diff
@@ -0,0 +1,664 @@
...
+         * @name gemini#fetchOrders
+         * @description watches information on multiple orders made by the user
+         * @see https://docs.gemini.com/websocket-api/#order-events
+         * @param {string|undefined} symbol unified market symbol of the market orders were made in
+         * @param {int|undefined} since the earliest time in ms to fetch orders for
+         * @param {int|undefined} limit the maximum number of  orde structures to retrieve
+         * @param {object} params extra parameters specific to the gemini api endpoint
+         * @returns {[object]} a list of [order structures]{@link https://docs.ccxt.com/en/latest/manual.html#order-structure}
+         */
+        const url = this.urls['api']['ws'] + '/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked';
```
Comment (carlosmiei) [inline js/pro/gemini.js:395]: @pcriadoperez `watchOrders` is throwing this network error: 
image


can you please take a look?

File: js/pro/gemini.js (hunk 1 above)
Comment (carlosmiei) [reply js/pro/gemini.js]: fixed

File: js/pro/gemini.js (hunk 3 above)
Comment (pcriadoperez) [reply js/pro/gemini.js:395]: I was getting the same issue from using `fetchMarkets` I opened this PR to solve the problem https://github.com/ccxt/ccxt/pull/15862

**Review by Blypto15 — COMMENTED** [review]
>

File: js/pro/gemini.js [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+        await this.loadMarkets ();
```
Comment (carlosmiei) [inline js/pro/gemini.js:396]: @pcriadoperez unfortunately even after the market's merge, I still get an error here:
image

and does not seem related to the markets because they're correctly loaded:
image


Can you please check this out?

Comment (pcriadoperez) [reply js/pro/gemini.js:396]: Good catch! I found the issue was that we weren't including headers in python websockets. I opened a seperate PR to address this issue: #16203

Comment (carlosmiei) [reply js/pro/gemini.js:396]: @pcriadoperez Unfortunately, even with https://github.com/ccxt/ccxt/pull/16203 merged I still get an error here. Weird thing is the error is different depending on the language:

Javascript:
```
CCXT v2.4.91
gemini.watchOrders ()
2022-12-31T15:37:51.286Z connecting to wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked
2022-12-31T15:37:51.691Z onError Unexpected server response: 400
2022-12-31T15:37:51.706Z onClose CloseEvent {
  [Symbol(kTarget)]: WebSocket {
    _events: [Object: null prototype] {
      open: [Function],
      message: [Function],
      error: [Function],
      close: [Function],
      ping: [Function: bound onPing],
      pong: [Function: bound onPong],
      upgrade: [Function: bound onUpgrade]
    },
    _eventsCount: 7,
```
```Python
p gemini watchOrders --verbose 
Python v3.10.8
CCXT v2.4.91
gemini.watchOrders()
2022-12-31T15:39:02.855Z connecting to wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked with timeout 10000 ms
2022-12-31T15:39:03.252Z NetworkError 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
2022-12-31T15:39:03.252Z on_error 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
Traceback (most recent call last):
  File "/Users/cjg/Git/ccxt7/ccxt/examples/py/cli.py", line 223, in <module>
    asyncio.run(main())
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/cjg/Git/ccxt7/ccxt/examples/py/cli.py", line 202, in main
    result = await method(*args)
  File "/Users/cjg/Git/ccxt7/ccxt/python/ccxt/pro/gemini.py", line 379, in watch_orders
    orders = await self.watch(url, messageHash, None, messageHash)
ccxt.base.errors.NetworkError: 400, message='Invalid response status', url=URL('wss://api.gemini.com/v1/order/events?eventTypeFilter=initial&eventTypeFilter=accepted&eventTypeFilter=rejected&eventTypeFilter=fill&eventTypeFilter=cancelled&eventTypeFilter=booked')
```

Comment (pcriadoperez) [reply js/pro/gemini.js:396]: strange, for me it's working for both js and python. Do you have any more info on the error to try and reproduce it?

Comment (carlosmiei) [reply js/pro/gemini.js:396]: @pcriadoperez yeah I have no idea, I tried several locations (uk, eua, etc) and got the exact same error

Comment (carlosmiei) [reply js/pro/gemini.js:396]: @pcriadoperez Great finding! 🙂 
Yeah the option seems a good idea, I will only change it slightly to be less dynamic. (basically `this[method]()` should be avoided IMO because it's hard to translate to typed languages)

**Comment by carlosmiei** [comment]
LGTM ✅ 

```
 p gemini watchTrades "ETH/USD" None 1          
Python v3.10.8
CCXT v2.5.33
gemini.watchTrades(ETH/USD,None,1)
[{'amount': 14.96809,
  'cost': 18700.8322842,
  'datetime': '2023-01-05T15:24:25.915Z',
  'fee': None,
  'fees': [],
  'id': '157890099255',
  'info': {'event_id': 157890099255,
           'price': '1249.38',
           'quantity': '14.96809',
           'side': 'sell',
           'symbol': 'ETHUSD',
           'timestamp': 1672932265915,
           'type': 'trade'},
  'order': None,
  'price': 1249.38,
  'side': 'sell',
  'symbol': 'ETH/USD',
  'takerOrMaker': None,
  'timestamp': 1672932265915,
  'type': None}]
  ```

```
p gemini watchOHLCV "ETH/USD" 1m None 5
Python v3.10.8
CCXT v2.5.33
gemini.watchOHLCV(ETH/USD,1m,None,5)
[[1672932900000, 1248.78, 1248.84, 1248.78, 1248.84, 0.02],
 [1672932960000, 1248.84, 1248.84, 1248.8, 1248.8, 0.00082],
 [1672933020000, 1248.8, 1248.8, 1248.8, 1248.8, 0.0],
 [1672933080000, 1248.8, 1248.8, 1248.4, 1248.4, 2.855323],
 [1672933140000, 1248.4, 1248.4, 1248.4, 1248.4, 0.0]]
 ```

```
Python v3.10.8
CCXT v2.5.33
gemini.watchOrderBook(ETH/USD,5)
{'asks': [[1249.46, 3.422541], [1249.84, 0.640172], [1249.85, 2.5], [1249.89, 2.5], [1249.9, 2.5], [1249.92, 2.5], [1249.93, 8.17], [1250.05, 2.513], [1250.11, 2.5], [1250.12, 3.187], [1250.17, 0.779949], [1250.32, 13.04], [1250.34, 1.5], [1250.41, 2.5], [1250.47, 2.5], [1250.48, 0.195889], [1250.5, 2.5], [1250.52, 2.5], [1250.95, 27.14], [1251.0, 3.202], [1251.12, 3.661674], [1251.24, 5.0], [1251.5, 52.18], [1251.71, 30.0], [1252.0, 51.232], [1252.23, 0.024094], [1252.27, 31.793], [1252.71, 0.004869], [1253.24, 0.064646], [1253.68, 54.86926], [1253.69, 191.016], [1253.81, 0.019075], [1254.23, 620.293], [1254.56, 0.01992], [1255.0, 6.39031], [1255.73, 0.004028], [1255.8, 130.09], [1256.38, 0.028023], [1256.74, 0.001], [1256.78, 0.001], [1256.79, 0.012321], [1256.83, 0.005014], [1257.14, 0.00754], [1257.42, 0.025671], [1257.6, 0.001], [1257.93, 0.001], [1258.03, 0.010433], [1258.31, 0.0818], [1258.4, 0.001], [1258.48, 0.023975], [1258.93, 109.01], [1259.4, 0.001], [1259.42, 0.001], [1259.5, 0.064323], [1259.8, 0.014418], [1259.86, 0.004841], [1260.0, 1.023212], [1260.1, 0.018981], [1260.26, 0.001], [1260.79, 0.001], [1261.0, 0.001], [1261.02, 0.001], [1261.74, 724.948], [1261.91, 0.018508], [1262.06, 0.019802], [1262.12, 0.001], [1262.24, 0.001], [1262.86, 0.001], [1262.94, 0.001], [1263.09, 0.001], [1263.15, 0.027873], [1263.9, 0.022712], [1264.0, 0.01225], [1264.55, 0.00754], [1264.57, 0.0217], [1264.58, 78.8], [1264.76, 0.023857], [1265.0, 0.021235], [1265.4, 0.010373], [1265.68, 0.081323], [1265.78, 0.064003], [1266.2, 803.467], [1266.26, 0.00498], [1266.43, 0.018886], [1266.52, 0.035257], [1266.57, 0.020251], [1266.81, 0.073154], [1266.84, 0.023606], [1267.07, 0.004815], [1267.12, 0.003992], [1267.96, 0.110486], [1269.61, 0.019684], [1269.95, 0.027724], [1270.0, 0.02], [1270.77, 0.025135], [1270.78, 0.004554], [1270.9, 0.001], [1271.07, 0.023738], [1271.26, 0.012181], [1271.31, 0.011452], [1271.65, 0.001], [1272.0, 0.00754], [1272.11, 0.063687], [1272.48, 0.033064], [1272.79, 0.018793], [1272.83, 0.010312], [1273.09, 0.08085], [1273.2, 0.0402], [1273.66, 31.908], [1273.68, 0.006289], [1273.75, 0.035635], [1274.31, 0.004791], [1275.0, 0.115726], [1275.63, 0.0049], [1275.77, 0.004941], [1275.88, 0.038945], [1276.01, 0.007999], [1276.02, 0.021086], [1276.51, 0.011751], [1276.79, 0.027575], [1276.86, 0.004533], [1277.2, 0.019567], [1277.27, 53.01], [1277.4, 0.02362], [1277.55, 10.0], [1277.78, 0.019877], [1278.0, 12.962959], [1278.46, 0.06337], [1278.55, 0.012111], [1278.62, 0.003956], [1278.8, 0.008649], [1278.93, 0.032344], [1279.0, 4.284647], [1279.18, 0.018698], [1279.49, 0.00754], [1279.59, 0.029717], [1280.0, 4.488282], [1280.28, 0.010252], [1280.29, 0.014174], [1280.3, 0.035263], [1280.5, 0.007019], [1280.56, 0.080378], [1280.7, 0.018237], [1280.85, 0.001], [1280.
```

**Review by carlosmiei — APPROVED** [review]
